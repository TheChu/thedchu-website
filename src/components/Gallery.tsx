import { graphql, useStaticQuery } from 'gatsby';
import { GatsbyImage } from 'gatsby-plugin-image';
import React, { FC, ReactElement } from 'react';

const Gallery: FC = (): ReactElement => {
  const {
    allMarkdownRemark: { edges },
  } = useStaticQuery(graphql`
    query {
      allMarkdownRemark(sort: { fields: frontmatter___date, order: DESC }) {
        edges {
          node {
            frontmatter {
              title
              date(formatString: "MMMM YYYY")
            }
            thumbnail {
              childImageSharp {
                gatsbyImageData(placeholder: BLURRED, quality: 90)
                original {
                  src
                }
              }
            }
          }
        }
      }
    }
  `);

  return (
    <section id="photos" className="bg-dark" aria-label="Photos">
      <div className="container-fluid p-0">
        <div className="row no-gutters">
          {edges.map(
            ({
              node: {
                frontmatter: { title, date },
                thumbnail: {
                  childImageSharp: {
                    gatsbyImageData,
                    original: { src },
                  },
                },
              },
            }) => (
              <div key={title} className="col-lg-3 col-sm-6">
                <a className="photos-box" href={src} aria-label={title}>
                  <GatsbyImage image={gatsbyImageData} alt={title} />
                  <div className="photos-box-caption">
                    <div className="project-category text-white-50">{date}</div>
                    <div className="project-name">{title}</div>
                  </div>
                </a>
              </div>
            )
          )}
        </div>
      </div>
    </section>
  );
};

export default Gallery;
