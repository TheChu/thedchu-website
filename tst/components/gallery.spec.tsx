import { render } from '@testing-library/react';
import { useStaticQuery } from 'gatsby';
import React from 'react';
import Gallery from '../../src/components/Gallery';

const mockEdge = (x: number) => ({
  node: {
    frontmatter: {
      title: `Test Title ${x}`,
      date: `June 200${x}`,
    },
    thumbnail: {
      childImageSharp: {
        gatsbyImageData: {
          layout: 'constrained',
          placeholder: {
            fallback: 'test-base64',
          },
          images: {
            fallback: {
              src: 'test-src',
              srcSet: 'test-srcSet',
              sizes: '100vw',
            },
            sources: [],
          },
          width: 1,
          height: 0.6666666666666666,
        },
        original: { src: 'test-src' },
      },
    },
  },
});
const edges = [mockEdge(1), mockEdge(2)];

(useStaticQuery as jest.Mock).mockReturnValue({
  allMarkdownRemark: {
    edges,
  },
});

describe('Gallery', () => {
  it('should match Gallery snapshot', () => {
    const { asFragment } = render(<Gallery />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the correct number of image links', () => {
    const { getAllByRole } = render(<Gallery />);
    expect(getAllByRole('link').length).toBe(2);
  });

  it('should render titles and dates', () => {
    const { getByRole } = render(<Gallery />);
    edges.forEach(
      ({
        node: {
          frontmatter: { title, date },
        },
      }) => {
        const link = getByRole('link', { name: title });
        expect(link).toBeVisible();
        expect(link).toHaveTextContent(date);
      }
    );
  });
});
