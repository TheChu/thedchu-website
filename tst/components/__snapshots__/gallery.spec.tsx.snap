// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery should match Gallery snapshot 1`] = `
<DocumentFragment>
  <section
    aria-label="photos"
    id="photos"
  >
    <div
      class="container-fluid p-0"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-lg-3 col-sm-6"
        >
          <a
            aria-label="Test Title 1"
            class="photos-box"
            href="test-src"
          >
            <div
              class="gatsby-image-wrapper gatsby-image-wrapper-constrained"
              data-gatsby-image-wrapper=""
              style="position: relative; overflow: hidden; display: inline-block; vertical-align: top;"
            >
              <div
                style="max-width: 1px; display: block;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  role="presentation"
                  src="data:image/svg+xml;charset=utf-8,%3Csvg height='0.6666666666666666' width='1' xmlns='http://www.w3.org/2000/svg' version='1.1'%3E%3C/svg%3E"
                  style="max-width: 100%; display: block; position: static;"
                />
              </div>
              <img
                alt=""
                aria-hidden="true"
                data-placeholder-image=""
                decoding="async"
                src="test-base64"
                style="height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              />
              <img
                alt="Test Title 1"
                data-gatsby-image-ssr=""
                data-main-image=""
                data-src="test-src"
                data-srcset="test-srcSet"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                style="height: 100%; left: 0px; position: absolute; top: 0px; transform: translateZ(0); transition: opacity 250ms linear; width: 100%; will-change: opacity; opacity: 0;"
              />
              <noscript />
              <script
                type="module"
              >
                const t="undefined"!=typeof HTMLImageElement&&"loading"in HTMLImageElement.prototype;if(t){const t=document.querySelectorAll("img[data-main-image]");for(let e of t){e.dataset.src&&(e.setAttribute("src",e.dataset.src),e.removeAttribute("data-src")),e.dataset.srcset&&(e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset"));const t=e.parentNode.querySelectorAll("source[data-srcset]");for(let e of t)e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset");e.complete&&(e.style.opacity=1)}}
              </script>
            </div>
            <div
              class="photos-box-caption"
            >
              <div
                class="project-category text-white-50"
              >
                June 2001
              </div>
              <div
                class="project-name"
              >
                Test Title 1
              </div>
            </div>
          </a>
        </div>
        <div
          class="col-lg-3 col-sm-6"
        >
          <a
            aria-label="Test Title 2"
            class="photos-box"
            href="test-src"
          >
            <div
              class="gatsby-image-wrapper gatsby-image-wrapper-constrained"
              data-gatsby-image-wrapper=""
              style="position: relative; overflow: hidden; display: inline-block; vertical-align: top;"
            >
              <div
                style="max-width: 1px; display: block;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  role="presentation"
                  src="data:image/svg+xml;charset=utf-8,%3Csvg height='0.6666666666666666' width='1' xmlns='http://www.w3.org/2000/svg' version='1.1'%3E%3C/svg%3E"
                  style="max-width: 100%; display: block; position: static;"
                />
              </div>
              <img
                alt=""
                aria-hidden="true"
                data-placeholder-image=""
                decoding="async"
                src="test-base64"
                style="height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              />
              <img
                alt="Test Title 2"
                data-gatsby-image-ssr=""
                data-main-image=""
                data-src="test-src"
                data-srcset="test-srcSet"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                style="height: 100%; left: 0px; position: absolute; top: 0px; transform: translateZ(0); transition: opacity 250ms linear; width: 100%; will-change: opacity; opacity: 0;"
              />
              <noscript />
              <script
                type="module"
              >
                const t="undefined"!=typeof HTMLImageElement&&"loading"in HTMLImageElement.prototype;if(t){const t=document.querySelectorAll("img[data-main-image]");for(let e of t){e.dataset.src&&(e.setAttribute("src",e.dataset.src),e.removeAttribute("data-src")),e.dataset.srcset&&(e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset"));const t=e.parentNode.querySelectorAll("source[data-srcset]");for(let e of t)e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset");e.complete&&(e.style.opacity=1)}}
              </script>
            </div>
            <div
              class="photos-box-caption"
            >
              <div
                class="project-category text-white-50"
              >
                June 2002
              </div>
              <div
                class="project-name"
              >
                Test Title 2
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
